src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1993,62 +2038,107 @@ SubqueryPushdownMultiPlanTree(Query *queryTree)
+	AttrNumber targetNo = 1;
+	List *subqueryTargetEntryList = NIL;
+	List *columnNamesList = NIL;
+	StringInfo rteName = makeStringInfo();
+	List *knownColumnList = NIL;
 
 	/* verify we can perform distributed planning on this query */
-	ErrorIfQueryNotSupported(queryTree);
-	ErrorIfSubqueryJoin(queryTree);
+	ErrorIfQueryNotSupported(queryTree, true);
 
 	/* extract qualifiers and verify we can plan for them */
 	qualifierList = QualifierList(queryTree->jointree);
 	ValidateClauseList(qualifierList);
 
 	/*
-	 * We disregard pulled subqueries. This changes order of range table list.
-	 * We do not allow subquery joins, so we will have only one range table
-	 * entry in range table list after dropping pulled subquery. For this reason,
-	 * here we are updating columns in the most outer query for where clause
-	 * list and target list accordingly.
+	 * We would be creating a new Query and pushing down top level query's
+	 * contents down to it. Join and filter clauses in higher level query would
+	 * be transferred to lower query. Therefore after this function we would
+	 * only have a single range table entry in the top level query. We need to
+	 * create a target list entry in lower query for each column reference in
+	 * upper level query's target list, and update varno/varattno attributes
+	 * of them accordingly.
 	 */
-	Assert(list_length(subqueryEntryList) == 1);
-
-	qualifierColumnList = pull_var_clause_default((Node *) qualifierList);
-	targetListColumnList = pull_var_clause_default((Node *) targetEntryList);
-
-	columnList = list_concat(qualifierColumnList, targetListColumnList);
+	columnList = pull_var_clause_default((Node *) targetEntryList);
 	foreach(columnCell, columnList)
 	{
+		TargetEntry *newTargetEntry = makeNode(TargetEntry);
+		ListCell *knownColumnCell = NULL;
+		StringInfo columnNameString = makeStringInfo();
 		Var *column = (Var *) lfirst(columnCell);
+		Var *foundColumn = NULL;
+		targetNo = 1;
+		foreach(knownColumnCell, knownColumnList)
+		{
+			Var *knownColumn = (Var *) lfirst(knownColumnCell);
+			if (knownColumn->varno == column->varno &&
+				knownColumn->varattno == column->varattno)
+			{
+				foundColumn = knownColumn;
+				break;
+			}
+			targetNo++;
+		}
+
+		if (foundColumn == NULL)
+		{
+			knownColumnList = lappend(knownColumnList, copyObject(column));
+			newTargetEntry->expr = (Expr *) copyObject(column);
+
+			appendStringInfo(columnNameString, WORKER_COLUMN_FORMAT,
+							 targetNo);
+			newTargetEntry->resname = columnNameString->data;
+
+			newTargetEntry->resjunk = false;
+			newTargetEntry->resno = targetNo;
+
+			subqueryTargetEntryList = lappend(subqueryTargetEntryList, newTargetEntry);
+			columnNamesList = lappend(columnNamesList, makeString(
+										  columnNameString->data));
+		}
+
 		column->varno = 1;
+		column->varattno = targetNo;
 	}
 
+	/* new query only has target entries and join tree */
+	queryCopy->targetList = subqueryTargetEntryList;
```

Comment:
Why are we using a copy of the original query? 
It makes really hard to follow the code given that we're actually trying to replace the query with a projection query. So, I'd suggest we make this intention clear such as only creating a new query object and getting the required fields (there may be other fields we should fill):

```C
	projectionQuery = makeNode(Query);
	projectionQuery->rtable = copyObject(queryTree->rtable);
	projectionQuery->jointree = copyObject(queryTree->jointree);
	projectionQuery->commandType = CMD_SELECT;
	projectionQuery->targetList = subqueryTargetEntryList;

```

And, I think this logic (ie., including the above column preparation) should live in another function given that it is completely separate what we're doing here.